services/src/datasets/in_memory.rs

```diff
@@ -231,6 +229,30 @@ impl DatasetProvider for HashMapDatasetDb {
             .cloned()
             .ok_or(error::Error::UnknownDatasetId)
     }
+
+    async fn provenance(
+        &self,
+        _session: &SimpleSession,
+        dataset: &DatasetId,
+    ) -> Result<ProvenanceOutput> {
+        match dataset {
+            DatasetId::Internal { dataset_id: _ } => self
+                .datasets
+                .iter()
+                .find(|d| d.id == *dataset)
+                .map(|d| ProvenanceOutput {
+                    dataset: d.id.clone(),
+                    provenance: d.provenance.clone(),
+                })
+                .ok_or(error::Error::UnknownDatasetId),
+            DatasetId::External(id) => {
+                self.dataset_provider(&SimpleSession::mock(), id.provider_id) // TODO: get correct session into dataset provider
```

Comment:
Why don't you use `_session`?

Reply:
adab347397faa698db32ae904b15882856f6dc56